Feature/qutech vsm driver

bergmansj opened the pull request:
Please use the following template for a pull request. (I will try, this is a first for me.)

Adds a qcodes driver for the QuTech VSM.
This pull request was requested by @duijedeurloo , but should probably be reviewed by
@AdriaanRol or @wvlothuizen 

The conditions below are probably not met because I wouldn't know how to perform
all these tests. Probably performing base SCPI testing of the VSM will be more effective.

In order for the pull request to be merged, the following conditions must be met:
- travis test suite passes
- all reasonable issues raised by codacy must be resolved 
- a positive review is required

Whenever possible the pull request should
- follow the PEP8 style guide 
- have tests for the code
- be well documented and contain comments

Tests are not mandatory as this is generally hard to make for instruments that interact with hardware.

AdriaanRol commented on pycqed/instrument_drivers/physical_instruments/QuTech_VSM_Module.py:
```diff
@@ -0,0 +1,267 @@
+        """
+        return self._socket.makefile().readline().rstrip()
+
+    def connect_message(self, idn_param='IDN', begin_time=None):
```
I think one should not overwrite a method like this for the reason stated. I agree with the general point but I would like to prevent any divergence form the QCoDeS base. As far as I know a proper logging solution is on the list for QCoDeS core.

bergmansj replied:
Fixed this by inheriting from SCPI (just like the AWG driver)

AdriaanRol commented on pycqed/instrument_drivers/physical_instruments/QuTech_VSM_Module.py:
```diff
@@ -0,0 +1,267 @@
+
+    def add_temperature_parameters(self):
+        # Command class temperature
+        self.add_parameter('setpoint',
```
It would be nice if this parameter has a unit I expect 'K' or "fancy symbol C"

bergmansj replied:
Fancy symbol it is :)

AdriaanRol commented on pycqed/instrument_drivers/physical_instruments/QuTech_VSM_Module.py:
```diff
@@ -0,0 +1,267 @@
+
+    def add_temperature_parameters(self):
+        # Command class temperature
+        self.add_parameter('setpoint',
```
Also a docstring would be nice (can be added by docstring="") , I don't know the difference between the setpoint and average temperature.

AdriaanRol commented on pycqed/instrument_drivers/physical_instruments/QuTech_VSM_Module.py:
```diff
@@ -0,0 +1,267 @@
+                           set_cmd='TEMPERATURE:SETPOINT {}',
+                           vals=vals.Numbers(min_value=10.0, max_value=55.0))
+        self.add_parameter('temperature_avg',
+                           get_cmd='TEMPERATURE?',
```
unit+ docstring

AdriaanRol commented on pycqed/instrument_drivers/physical_instruments/QuTech_VSM_Module.py:
```diff
@@ -0,0 +1,267 @@
+        for mod in self.modules:
+            mod_name = 'mod{m}'.format(m=mod)
+            mod_scpi = 'MODULE{m}'.format(m=mod)
+            self.add_parameter(mod_name + '_temperature_avg',
```
Even more temperatures here. 
Again some docstring that tells me which one it is would be useful.

AdriaanRol commented on pycqed/instrument_drivers/physical_instruments/QuTech_VSM_Module.py:
```diff
@@ -0,0 +1,267 @@
+        self.add_parameter('marker_source',
+                           set_cmd='MARKER:SOURCE {}',
+                           vals=vals.Enum('int', 'ext'))
+        self.add_parameter('marker_state',
```
The difference between marker state and marker_source is unclear to me. Could you add a docstring?

AdriaanRol commented on pycqed/instrument_drivers/physical_instruments/QuTech_VSM_Module.py:
```diff
@@ -0,0 +1,267 @@
+                           set_cmd='MARKER:STATE {}',
+                           vals=vals.OnOff())
+
+        for mod in self.modules:
```
You seem to both add the parameter once and after that add it on a per module basis. First what is the difference between the two and why do you add it twice?

AdriaanRol commented on pycqed/instrument_drivers/physical_instruments/QuTech_VSM_Module.py:
```diff
@@ -0,0 +1,267 @@
+        for channel in self.channels:
+            ch_name = '_ch{c}'.format(c=channel)
+            ch_scpi = ':CHANNEL{c}'.format(c=channel)
+            self.add_parameter('qubit' + ch_name + '_description',
```
what is qubit description used for? Is this used for the monitor/display? Also please add a docstring.

AdriaanRol commented on pycqed/instrument_drivers/physical_instruments/QuTech_VSM_Module.py:
```diff
@@ -0,0 +1,267 @@
+                               get_cmd='QUBIT'+ch_scpi+':DESCRIPTION?',
+                               set_cmd='QUBIT'+ch_scpi+':DESCRIPTION {}',
+                               vals=vals.Strings())
+            self.add_parameter('qubit' + ch_name + '_frequency',
```
What units is this in? It should be in Hz (we do everything in SI). If the device thinks e.g. in GHz then please add a conversion. 

Also could you add a docstring on what this is used for? I think it relates to what calibration data is used.

AdriaanRol commented on pycqed/instrument_drivers/physical_instruments/QuTech_VSM_Module.py:
```diff
@@ -0,0 +1,267 @@
+                               get_cmd='QUBIT'+ch_scpi+':FREQUENCY?',
+                               set_cmd='QUBIT'+ch_scpi+':FREQUENCY {}',
+                               vals=vals.Numbers())
+            self.add_parameter('qubit' + ch_name + '_led_color',
```
what kind of strings are accepted here? basic color names? hex values? Please add a docstring.

AdriaanRol commented on pycqed/instrument_drivers/physical_instruments/QuTech_VSM_Module.py:
```diff
@@ -0,0 +1,267 @@
+
+    def add_parameters(self):
+        # VSM properties
+
```
I like the grouping of add parameters here.

AdriaanRol commented on pycqed/instrument_drivers/physical_instruments/QuTech_VSM_Module.py:
```diff
@@ -0,0 +1,267 @@
+                # All channels and modules (no getter)
+                var_name = '_{p}_{a}_raw'.format(p=pulse.lower(), a=ap.lower())
+                var_scpi = ':{p}:{a}:RAW'.format(p=pulse, a=ap)
+                self.add_parameter('calibration' + var_name,
```
What is a calibration "Int"? and what does it do? Please add a docstring (and if it refers the state it should be a Bool).

AdriaanRol commented on pycqed/instrument_drivers/physical_instruments/QuTech_VSM_Module.py:
```diff
@@ -0,0 +1,267 @@
+                                           vals=vals.Ints())
+
+
+if __name__ == '__main__':
```
I guess the __name__=='__main__': is there for testing purposes. We generally use drivers as a module and as such this bottom part should be removed to keep the code clean. 
Adding an example to the examples folder with this code snippet would be useful but not required.

AdriaanRol commented on pycqed/instrument_drivers/physical_instruments/QuTech_VSM_Module.py:
```diff
@@ -0,0 +1,267 @@
+
+        # Command class calibration
+        for pulse in ('GAUSSIAN', 'DERIVATIVE'):
+            for ap in ('ATTENUATION', 'PHASE'):
```
I can see the calibration paramter but I can find nowhere where we can set the channel specific attenuation/phases. Also (once these are added) what units are these in? dB? raw dac values?

AdriaanRol commented:
Hi @bergmansj my main review comments can be summarised in the following points: 
- There are no docstrings and units making it very hard to understand how things work. 
- You seem to be adding every parameter once globally and then again for each module. This seems like a bug unless I misunderstand. 
- The parameters to set the channel specific Attenuation and Phase do not seem to be added. 

I would greatly appreciate it if you could resolve these issues.

bergmansj commented:
Changes:

- Inherit from SCPI, like the AWG driver.
- Added parameter docstrings and extended module documentation.
- Where applicable added units.
- Removed parameters from the driver that have no prio yet.
- Removed commands for temperature setpoints (of the heating PID controllers)
  because we probably don't want to allow for changing those.
- No validators for getters. (why are there validators on the driver side at all?)
- Removed `__main__` block from module.
- Calibration commands for each qubit separately, to keep things simple. This
  can be extended later when requested.
- During development, internal VSM logs can be viewed by pointing a
  webbrowser to port 5000 of the VSM, example: http://192.168.0.10:5000/
  For normal use, this will probably be restricted/disabled.
- Some code-yoga to adhere to PEP8 :)

bergmansj commented:
Sorry for not responding earlier this week. I processed all your comments/suggestions. Please check them out.

NielsBultink commented:
Here, my first impressions.
several issues with the driver and functioning of the VSM as it is now:

**blocking:**
- stability:
when reading out all commands: for instance VSM.print_human_readable_snapshot() or station.add_component(VSM), the device often loses communication. Only way to solve this is power cycling. 

- individual marker_source settings lead to a crash
![image](https://user-images.githubusercontent.com/12341001/33037055-a2f39bc2-ce30-11e7-8e80-5ff816db90e3.png)

**not blocking but major drawback:**
- all integer and float get commands come back as strings: temperatures, attenuations, phases

**not implemented:**
- as it seems,orthogonalization of phase and attenuation settings using the on-board calibration files. Also, rescaling of the amplitudes a float between 0.x and 1 is not implemented

bergmansj commented:
Thanks for looking into the driver code. 

- stability:
when reading out all commands: for instance VSM.print_human_readable_snapshot() or station.add_component(VSM), the device often loses communication. Only way to solve this is power cycling. 

_These are probably methods added by the Qcodes magic. I didn't put them into the driver Too advanced for the VSM at this point maybe? what SCPI commands are involved?_ 

- individual marker_source settings lead to a crash

_The marker _source_ can only be set for a complete module, not er channel._

- all integer and float get commands come back as strings: temperatures, attenuations, phases

_Probably a result of me misunderstanding the Qcodes driver magic that is needed. The SCPI protocol is string based, so everything has to be transported as strings. Feel free to correct the driver code for this._

- as it seems,orthogonalization of phase and attenuation settings using the on-board calibration files. Also, rescaling of the amplitudes a float between 0.x and 1 is not implemented

_What do you mean by this? The calibration files are not implemented yet, because it basically is syntactic 'linarization' over the control of the attenuation and phase DACs. I even removed the preparatory code from the driver to keep things better testable._

AdriaanRol commented:
@bergmansj 
> Thanks for looking into the driver code.

>> stability:
>> when reading out all commands: for instance VSM.print_human_readable_snapshot() or station.add_component(VSM), the device often loses communication. Only way to solve this is power cycling.

> These are probably methods added by the Qcodes magic. I didn't put them into the driver Too advanced for the VSM at this point maybe? what SCPI commands are involved?

The magic here is almost non existent, the station.add_component reads out all parameter (calls `.get` on every parameter) before adding it to a list of components. Something as basic as reading out parameters should not lead to a crash. The SCPI commands involved are whatever get commands added to the driver. 

>> individual marker_source settings lead to a crash

> The marker source can only be set for a complete module, not er channel.

If an action is not supported in the hardware, the driver should not allow us to do this. More importantly, if we try to do something that is not allowed it should raise an exception, not crash the VSM. 

>> all integer and float get commands come back as strings: temperatures, attenuations, phases

> Probably a result of me misunderstanding the Qcodes driver magic that is needed. The SCPI protocol is string based, so everything has to be transported as strings. Feel free to correct the driver code for this.

It is as simple as adding a `get_parser=float` (in case of a float) as an argument to the add_parameter. I have no problem adding this but would prefer not to touch the driver until it has been properly tested. 


>>  as it seems,orthogonalization of phase and attenuation settings using the on-board calibration files. Also, rescaling of the amplitudes a float between 0.x and 1 is not implemented

> What do you mean by this? The calibration files are not implemented yet, because it basically is syntactic 'linarization' over the control of the attenuation and phase DACs. I even removed the preparatory code from the driver to keep things better testable.

We were under the impression that the VSM was completely done and that the driver would support the orthonormalized/linear mode. If this is not the case I guess we should discuss with @duijedeurloo and Jeremy correct? 

Best, 

Niels and Adriaan

NielsBultink commented:
As discussed with Jeroen, here a simple example code to make the unit crash:
In this, example I have made to attempts to avoid crashes:
- add a sleep time between each setting, still experimenting with making this longer
- add parameter assertion after each parameter setting

![image](https://user-images.githubusercontent.com/12341001/33074491-b12cf440-cec6-11e7-8c6f-dfaff5f344ca.png)

crashing throws the following error:
![image](https://user-images.githubusercontent.com/12341001/33074566-eb88a648-cec6-11e7-92a0-00e6071c386d.png)

bergmansj commented:
@NielsBultink I made some fixes so that the code as pasted above (next time copy-pastable text please) does no longer crash the message queue.

NielsBultink commented:
ah, interesting. Is this already pushed to the feature branch?

2017-11-21 16:31 GMT+01:00 Jeroen Bergmans <notifications@github.com>:

> @NielsBultink <https://github.com/nielsbultink> I made some fixes so that
> the code as pasted above (next time copy-pastable text please) does no
> longer crash the message queue.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/DiCarloLab-Delft/PycQED_py3/pull/396#issuecomment-346062849>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ALxPCQXzfV8Ur7wxO0XHEhEoYH1ksTy3ks5s4uzAgaJpZM4QTlZg>
> .
>

AdriaanRol commented:
![image](https://user-images.githubusercontent.com/6142932/33084434-4381c81e-cee2-11e7-8d2d-970821974c27.png)

Looks like it.

NielsBultink commented:
not to me

> On 21 Nov 2017, at 17:34, Adriaan <notifications@github.com> wrote:
> 
>  <https://user-images.githubusercontent.com/6142932/33084434-4381c81e-cee2-11e7-8d2d-970821974c27.png>
> Looks like it.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub <https://github.com/DiCarloLab-Delft/PycQED_py3/pull/396#issuecomment-346084040>, or mute the thread <https://github.com/notifications/unsubscribe-auth/ALxPCRLcD1-ZnRij8SiPcudQgLxx-b6Qks5s4vurgaJpZM4QTlZg>.
>

AdriaanRol commented:
@NielsBultink have all the changes made while testing today been committed to this branch? 
If so we can merge.
